src/cfnlint/rules/resources/s3/BucketPolicy.py

```diff
@@ -0,0 +1,238 @@
+        """Init"""
+        super(BucketPolicy, self).__init__()
+        self.resource_property_types = ['AWS::S3::BucketPolicy']
+
+    # pylint: disable=W0613
+    def check_value(self, value, path, start_mark, end_mark, bucket_resources, custom_resources):
+        """Count them up """
+
+        if isinstance(value, six.string_types):
+            try:
+                value = convert_dict(json.loads(value), start_mark, end_mark)
+            except Exception as ex:  # pylint: disable=W0703,W0612
+                return []
+
+        if isinstance(value, dict):
+            return self._check_policy(value, path[:], bucket_resources, custom_resources)
+
+        return []
+
+    def _check_policy(self, policy, path, bucket_resources, custom_resources):
+        """ Check the entire policy """
+        matches = []
+
+        key = 'Statement'
+        for statements, statements_path in policy.get_safe(key, default=[], path=path[:]):
+            if isinstance(statements, list):
+                for statement, statement_path in statements.items_safe(path=statements_path[:]):
+                    matches.extend(
+                        self._check_statement(
+                            statement, statement_path[:],
+                            bucket_resources, custom_resources))
+
+        return matches
+
+    def _check_statement(self, value, path, bucket_resources, custom_resources):
+        """ Check each statement"""
+        matches = []
+
+        actions = value.get('Action')
+        if not isinstance(actions, (list, six.string_types)):
+            return matches
+        resources = value.get('Resource')
+        if not isinstance(resources, (list, six.string_types, dict)):
+            return matches
+
+        if isinstance(actions, six.string_types):
+            actions = [actions]
+        if isinstance(resources, (six.string_types, dict)):
+            resources = [resources]
+
+        found_object_actions, found_bucket_actions, actions_results_conclusive = self._get_statement_actions(actions)
+        found_object_resources, found_bucket_resources, bucket_results_conclusive = self._get_statement_resources(
+            resources, bucket_resources, custom_resources)
+        if not actions_results_conclusive or not bucket_results_conclusive:
+            return matches
+
+        if (not found_object_resources) and found_object_actions:
+            message = 'S3 Bucket Policy has object actions and no specified object resource'
+            matches.append(RuleMatch(path[:] + ['Resource'], message))
+        elif not found_bucket_resources and found_bucket_actions:
+            message = 'S3 Bucket Policy has bucket actions and no specified bucket resource'
+            matches.append(RuleMatch(path[:] + ['Resource'], message))
+
+        return matches
+
+    def _get_statement_actions(self, actions):
+        """ Breaks down the actions in a statement to each category """
+        object_actions = [
```

Comment:
Can't we grab this data from the IAM policies? (from the experimental rule)

Reply:
Looking through that policy definition it doesn't look like it distinguishes between objects and the bucket.  So I don't believe so.  I wish I could though.

Reply:
@kddejong thoughts on something like
```
object_actions = list(filter(re.compile("s3:.*Object").match, list_of_all_policies)) + [
            's3:AbortMultipartUpload',
            's3:ListMultipartUploadParts',
            's3:BypassGovernanceRetention',
]
```

Reply:
Let me compare the results.

Reply:
Updated to be more dynamic.